### ISSUE TITLE
We need to chain multiple Actions with delays (sending multiple Emails at some intervals)

### ISSUE THREAD

**Issue by pitus** [mautic/mautic#266]
This is a Feature Request more than a Bug report.

We need to send several emails spaced at a specific delay/interval such as 7 days since the subscription and currently, there's no way to chain multiple Email (or any other) Actions together. We're forced to use Decisions between them, but there are no decisions to make in our case so we would have to use the negative (red) outcome of a fake decision so that it's always taken ... and triggers another email action.
The problem is that the delay time have to be all based on the initial action rather than on the previous one so it's a bit of a pain to define and update...

It would make a lot of sense to allow multiple Actions to be chained/linked to previous Action (it's like a Procedure in programming parlor, multiple commands treated as one) - hope you'll consider this.

-Piotr

**Comment by alanhartless** [comment]
Thanks for suggesting this Piotr.  

Can you give me an example of your campaign? 

If you are just sending interval emails in the campaign, you should be able to have them all as the top level set at the different intervals (so each action would be standalone without being connected into anything).  From a UI standpoint, I know the interval won't show automatically as it does when connected (planing to fix this) but it should work without using a decision. 

Thanks,
Alan

**Comment by pitus** [comment]
Thanks for your suggestion of adding all Email Actions just below the trigger (decision), this might actually work (haven't thought of that before - as a programmer, I tend to think of multiple Actions as a sequence not in parallel ;)

The only difference between this way of handling it, with delayed actions, is that it may be easier to define them as chained where the Delay time would be measured counting from the previous step/action in the chain rather than from the initial trigger...

Say I want to run a training campaign where the user subscribes to a set of training emails, we want them to be sent one step per week since the day they've subscribed. Yes, we could just add all of them at the top and just use different delays for each, but if they could be defined as Send Email 1 after 7 days (linked to the top), then another action of Send Email 2 after 7 days (linked to the Email 1 and counting from it rather than 2 weeks from the top) ... then it would be easier to manage and visually comprehend.

It's not critical, now that we can work around this limitation, but I don't know where/how to designate an issue as a feature request vs bug ... ?

**Comment by MaxOnTheHill** [comment]
@pitus your suggestion makes good sense and is something I would think many would expect to work in that manner.  If it can't work this way currently, then expect it is a feature request ... but is one that could lead to many bug/issue submissions in the future if the action if not added to the core, as is seen to be an expected action.    +1 from me ;)

**Comment by alanhartless** [comment]
@pitus Unless there is a bug, the parallel actions at the root level (i.e. those starting the campaign; not connected from a decision) with intervals set should use the date/time the lead was added to the campaign to schedule events against. 

For example, I created a campaign with three root level send email actions:
![](http://cl.ly/image/0c1Z140g3f0x/Image%202015-04-06%20at%2020%3A38%3A02.png)

I added my lead to the campaign and I ended up with the events scheduled as per the interval for the lead:
![](http://cl.ly/image/061d2r0d3a3Q/Image%202015-04-06%20at%2020%3A37%3A01.png)

Are you finding the events are not getting scheduled or are they not getting executed?

Thanks,
Alan

**Comment by argentin** [comment]
@alanhartless they are not getting executed! try with 1 hour delay and see what happens... thanks for your support cc @MaxOnTheHill

**Comment by alanhartless** [comment]
@argentin Just to verify; are you running the command to execute them via a cron job? See https://www.mautic.org/docs/campaigns/campaign_builder.html#non-action-decision-points.

**Comment by alanhartless** [comment]
(Need to update the doc; it talks about the non-action decision points there but the same command executes scheduled events as well).

**Comment by argentin** [comment]
@MarceloPedra will verify this in our mautic installation

**Comment by MarceloPedra** [comment]
Hi @argentin + @alanhartless , what's the recommended frequency to run the cronjob?

**Comment by alanhartless** [comment]
There isn't really a recommendation. It depends on your setup. If most of your campaigns (also think of future campaigns) have intervals of hours, then I would set it for every hour. If days, then every day. If minutes, then every few minutes.  

Thanks,
Alan

**Comment by pitus** [comment]
What I find most confusing in the "non-action" outcome is that the documentation doesn't really state when it is being executed and what is the recommended cron job configuration for most scenarios.

For example, I have a campaign where a Lead may open an email any time (depending when they get a chance to look at the email or ... when they've joined the campaign (an email was sent to them some time later than the campaign started) and we want to check if they've opened it within say 7 days since the email was **sent** to them (not since the campaign has started). If they do, fine - take the green path... unfortunately the red (non-action) path is taken as well, leading to an additional, unnecessary action to be taken (since they've actually opened their email within 7 days). Why is the non-action being executed in such case? Shouldn't it be blocked when the user opens the email? Otherwise, it makes absolutely no sense to call it "email NOT opened"... or "NO email opened".
Next, when does the system "decide" that the email has NOT been opened yet? There's no place in the decision to define the time-frame, so I'm assuming that any cron job that runs campaigns trigger will cause the red action to be taken (in our case, it's ran every few minutes so the non-action is taken at most few minutes after the email has been sent - complete nonsense)... We can't be setting up separate cron jobs for each campaign just to specify when the non-action is to be taken, that delay should be specified in the decision itself! Especially so, when we can't tell when the email has actually been sent (the Lead has joined the campaign few days after it has started - we still want to give him 7 days to open it rather than 7-few it took them to join... based on a cron ran at a given date/time to trigger the non-action.

Finally, supposedly the non-action has been triggered and then the Lead has opened the email (after it's been deemed un-opened) ... What should happen? Currently the green action is taken even though the Lead should have been flagged as non-opened... Perhaps there should be another option for that (mutual exclusion of yes/no outcomes) ?

**Comment by alanhartless** [comment]
Hi Pitus, 

As mentioned in #264, the triggering of the non-action events are not functioning as intended. I just submitted PR #273 if you would be willing to help me test it?  Actions attached to non-action decision points (red) are not supposed to be executed unless the lead did not complete the action decision point (green).  The bug was that there was a DQL query that ended up executing the actions associated with non-decision points for more leads than intended. Hopefully #273 fixes that.

For the last question, I think you are are right that there is not a hard stop to prevent the "green" path if the lead has already traversed the "red" path. There is a stop for the other way around though (can't go down the red if they are already down the green). I'll look at that as well. I'm assuming the assumed outcome is that the lead not be allowed to transverse the "green" path if they have already started down the "red" path?

Thanks,
Alan

**Comment by pitus** [comment]
The mutual exclusion of green vs red should probably be configurable per each decision. There may be cases where the red path has been already taken and then the Lead has opened the email and probably should trigger the green action - it depends on the campaign so why not let the campaign creator decide if green should be taken when red has already triggered?

**Comment by alanhartless** [comment]
Allowing the decision to determine if a "green" path should be taken or not if the lead has already started down the "red" (via a config option in the decision form) will take some logic changes in the code so should probably be requested as a new feature request issue.

Thanks!
Alan

**Comment by pitus** [comment]
How do I request a new feature in github?

**Comment by dbhurley** [comment]
Just create a new issue, someone here will add a "feature" tag to it. :+1:

**Comment by pitus** [comment]
One more detail. You said:
> Actions attached to non-action decision points (red) are not supposed to be executed unless the lead did not complete the action decision point (green).

And the https://www.mautic.org/docs/campaigns/campaign_builder.html#non-action-decision-points page has the following statement (which is probably the one referred to above as needing a clarification):
> That command will find and execute scheduled events and/or non-action events that have passed the `specified time frame.`

The question is, when does the system decide that the action has **NOT been taken** ? There must be some way of specifying how long to wait until deciding it's not going to be taken anymore, let's consider it a NO ... but there's no place to define this `specified time frame` for a non-action.

If the cron job `php /path/to/mautic/app/console mautic:campaigns:trigger --env=prod` is run every few minutes then will it trigger all non-actions at each execution? Documentation suggests that it'll *check* if it needs to be taken, as if it knew what kind of timeframe to look for when making that decision?

**Comment by alanhartless** [comment]
Hi Pitus,

The specified non-action time is the "delay" in the connected action. So if you connect an action with a 1 day delay into the "non-action" or red path of a decision, if the lead does not execute the decision within 1 day, the non-action path will be taken. We need to find a way to make that more clear in the UI.  The problem is that the same action is shared between the two and won't know what the user is going to connect it into until after the form is saved.  But we could probably reword the Add a delay label to help clarify.

As mentioned already a couple times in this GH issue and the other related discussion, the execution of all non-action decision is a bug and is one of the fixes included in https://github.com/mautic/mautic/pull/273.

Thanks,
Alan

**Comment by pitus** [comment]
Thanks for clarifying this - that the Delay for a Non-Action (red) output of a Decision is define in the Action linked to that red output (if I'm reading it correctly ;)

In fact, that's what we tried initially assuming it will work as expected but since it didn't (due to the #273 bug), we tried to figure it out with various other possibilities such as controlling the cron job timing...

I'll download the latest fixes and try them out as we will need some of this in our upcoming campaign.

**Comment by alanhartless** [comment]
Yea I'm afraid due to the way the bug triggered itself, manipulating the cron job wouldn't help. We are looking at 1.0.3 real soon because of the campaign bugs discovered. The latest fixes will only be available through #273 until we merge it. But you can snag the files from the PR and apply them https://github.com/mautic/mautic/pull/273/files.  Definitely let us know the results if you do apply the changes!

**Comment by alanhartless** [comment]
I think this one is resolved with latest releases and the face that events can be added at root level with a configured delay.  If the issues are still there; let's open a new clean issue as this one got mixed with several things.

Thanks!
Alan
